### PR TITLE
Fix XR Body crouching and climbing

### PR DESCRIPTION
**Pull request by Malcolmnixon**
The XRBodyModifier3D node has a bone-update option of "Rotation Only" which is intended to preserve the lengths of bones. This is achieved by only setting the bone rotation rather than applying both rotation and position.

Unfortunately this option also skipped setting the "Hips" bone position which is the first actual bone of the body. As a result the players hips are fixed at the height of the T-Posed character.

The fix is to always set the Hips bone position and rotation to position the character, and then all child bones are updated as configured by the bone-update setting.

image

The following video demonstrates how this change affects character models performing different actions which affect vertical height:

https://github.com/godotengine/godot/assets/1863707/b455f77d-f8f7-4288-ac03-e2f3c8a6a650

### CODE REVIEW

File: scene/3d/xr_body_modifier_3d.cpp [hunk 1]
```diff
@@ -353,8 +353,9 @@ void XRBodyModifier3D::_update_skeleton() {
 		const int parent_joint = joints[joint].parent_joint;
 		const Transform3D relative_transform = inv_transforms[parent_joint] * transforms[joint];
 
-		// Update the bone position if enabled by update mode.
-		if (bone_update == BONE_UPDATE_FULL) {
+		// Update the bone position if enabled by update mode, or if the joint is the hips to allow
```
Comment (AThousandShips) [inline scene/3d/xr_body_modifier_3d.cpp:356]:
```suggestion
		// Update the bone position if enabled by update mode, or if the joint is the hips, to allow
```

**Comment by TokageItLab** [comment]
How is this handled with respect to root bone?

Also, not directly related to this, but for about the position value, probably it needs to be multiplied by the Skeleton3D.motion_scale.

**Comment by dsnopek** [comment]
How is the root bone "defined"? If it's defined at the bottom center of the avatar (or something similar), then I think positioning the hip bone makes sense. Since all bone positions are relative (except the root itself), that means the hip would be positioned relative to the root, and so we shouldn't have issues with that allowing the player to move through walls: it should be mostly Y position, and maybe some small X/Z variations (I guess as far as you can shift your hips over your center).

**Comment by TokageItLab** [comment]
I am not familiar with the specs of the XR controller, but if the XR tracker does not have the concept of root, then there should be an option for the movement of the XZ axis of the Hips that can be transferred to the Root.

At this point, whether or not the root_bone is actually moved should also be an option. Perhaps we should be able to specify a Modifier for the RootMotionView.

Well these may be a bit off-topic, but we need to consider that the option to use Hips or specify Root is missing and if it is added it may be included in the BONE_UPDATE enum.

--

By the way, regarding motion_scale, since GodotHumanoid has the waist height as motion_scale when retargeting, it is necessary to multiply the tracker's movement delta value by motion_scale.

In other words, the movement by GodotHumanoid's set_bone_pose_position() without multiplication is based on the assumption that the model has a hips height of 1 meter. Since the movement of a model with a hips height of 0.5 meter can be simply estimated to be a movement of 0.5 meter, the model's slip can be reduced by multiplying by the model's hips height.

**Comment by Malcolmnixon** [comment]
> How is the root bone "defined"? If it's defined at the bottom center of the avatar (or something similar), then I think positioning the hip bone makes sense. Since all bone positions are relative (except the root itself), that means the hip would be positioned relative to the root, and so we shouldn't have issues with that allowing the player to move through walls: it should be mostly Y position, and maybe some small X/Z variations (I guess as far as you can shift your hips over your center).

All the models I've put through the Godot Humanoid Skeleton Retargeting have the Root bone at ground level under the hips. The SkeletonProfileHumanoid has the "Root" at 0,0,0 and any minor positional issues may be automatically handled by the "Rest Fixer" or the "Fix Silhouette" features.

**Comment by Malcolmnixon** [comment]
> By the way, regarding motion_scale, since GodotHumanoid has the waist height as motion_scale when retargeting, it is necessary to multiply the tracker's movement delta value by motion_scale.

The trackers don't provide movement deltas - only the tracked absolute positions and rotations.

There appear to be two settings of interest with regards to scaling:
- `Skeleton3D.motion_scale` - useful for fitting the tracked player/actor joints to the size of the avatar
- `XRServer.world_scale` - useful for fitting the avatar to the world

I believe the correct set of actions is:
1. Apply the `Skeleton3D.motion_scale` to the set_bone_pose_position position parameter
2. Scale the XRBodyModifier3D transform by `XRServer.world_scale`

The intended result would be that differences in player/actor size could be adjusted to fit the size of the avatars skeleton, and then the entire avatar - mesh and skeleton would be scaled to fit the world (assuming the avatar is a child of the XRBodyModifier3D).

**Comment by TokageItLab** [comment]
> The trackers don't provide movement deltas - only the tracked absolute positions and rotations.

Then, I think the delta needs to be calculated on the XRBodyModifier3D in the same way as AnimationMixer for root_motion. It will be possible to follow up later.

**Review by dsnopek — APPROVED** [review]
@Malcolmnixon: 

> All the models I've put through the Godot Humanoid Skeleton Retargeting have the Root bone at ground level under the hips. The SkeletonProfileHumanoid has the "Root" at 0,0,0 and any minor positional issues may be automatically handled by the "Rest Fixer" or the "Fix Silhouette" features.

Ok, then this change makes sense to me conceptually! I haven't tested it, though.

**Comment by Malcolmnixon** [comment]
This PR attempts to resolve the scaling issue by:
1. Applying the `Skeleton3D.motion_scale` to the incoming joint positions
2. Applying the `XRServer.world_scale` to the XRBodyModifier3D transform (which includes the scaled root joint)

This gives the developer the ability to modify `Skeleton3D.motion_scale` to better match the tracking data to the skeleton, and then adjust `XRServer.world_scale` to uniformly scale the entire avatar against the world.

The following video demonstrates playing with these effects:

https://github.com/godotengine/godot/assets/1863707/9dad2b5a-316a-4fd1-9daf-8a2410331bfc

File: scene/3d/xr_body_modifier_3d.cpp [hunk 2]
```diff
@@ -311,18 +311,22 @@ void XRBodyModifier3D::_update_skeleton() {
 		return;
 	}
 
-	// Read the relevant tracking data.
+	// Get the world and skeleton scale
```
Comment (AThousandShips) [inline scene/3d/xr_body_modifier_3d.cpp:314]:
```suggestion
	// Get the world and skeleton scale.
```

File: scene/3d/xr_body_modifier_3d.cpp [hunk 3]
```diff
@@ -353,17 +357,20 @@ void XRBodyModifier3D::_update_skeleton() {
...
 		// Always update the bone rotation.
 		skeleton->set_bone_pose_rotation(joints[joint].bone, Quaternion(relative_transform.basis));
 	}
 
-	// Transform to the skeleton pose.
-	set_transform(transforms[XRBodyTracker::JOINT_ROOT]);
+	// Transform to the tracking data root pose. This also applies the XR world-scale to allow
+	// scaling the avatars mesh and skeleton appropriately (if they are child nodes).
+	set_transform(
+			transforms[XRBodyTracker::JOINT_ROOT].scaled(Vector3(ws, ws, ws)));
```
Comment (AThousandShips) [inline scene/3d/xr_body_modifier_3d.cpp:373]:
```suggestion
			transforms[XRBodyTracker::JOINT_ROOT] * ws);
```
Should be equivalent

Comment (Malcolmnixon) [reply scene/3d/xr_body_modifier_3d.cpp]: Confirmed it is - nice catch!

**Comment by Malcolmnixon** [comment]
@BastiaanOlij Could you check the updated documentation with regards to node scaling and positioning. It's similar to the proposed XRHandModifier3D changes.

**Review by BastiaanOlij — APPROVED** [review]
Looks good Malcolm.

We could (ab)use this opportunity to also mark this class as experimental. Best to get that out of the way before 4.3 releases :)

**Comment by Malcolmnixon** [comment]
> We could (ab)use this opportunity to also mark this class as experimental. Best to get that out of the way before 4.3 releases :)

Switched XRBodyTracker, XRBodyModifier3D, XRFaceTracker, and XRFaceModifier3D to experimental.

**Review by TokageItLab — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
